### ISSUE TITLE
Build steam-dos as an extension for Steam

### ISSUE THREAD

**Issue by LeandroStanger** [flathub/com.valvesoftware.Steam#394]
https://github.com/dreamer/steam-dos

**Comment by LeandroStanger** [comment]
@dreamer

**Comment by dreamer** [comment]
Hello! :) I think some work on this was already started by @gasinvein, but I don't know if it is being tracked in some issue already. Also, I think we are waiting for ValveSoftware/steam-for-linux#6310 to be released (although it's not a blocker for flathub extension if I understand correctly).

**Comment by gasinvein** [comment]
Yeah I have a working manifest for building steam-dos extension [here](https://github.com/gasinvein/flatpak-apps-staging/tree/com.valvesoftware.Steam.CompatibilityTool.steam-dos/master).
ValveSoftware/steam-for-linux#6310 isn't a blocker indeed, but the extension support is added only on the beta branch of this Steam flatpak, so the steam-dos extension would be only available from flathub-beta, too.
Also I know @dreamer is planning to rename steam-dos projects, and this actually is kind of blocker since changing flatpak IDs may be tricky.

**Comment by dreamer** [comment]
Yes, plan is to rename steam-dos project to Boxtron - hopefully by the end of next week (but the date is not set in stone).

In the parallel [Luxtorpeda](https://github.com/dreamer/Luxtorpeda) is being developed - it's somewhat functional, name is final, but no official release has happened yet. I think Luxtorpeda might be easier to package as an extension, as it is single Rust binary with no special dependencies.

**Comment by gasinvein** [comment]
@dreamer If the name Boxtron if already final, we can just preventively give the extension ID of `com.valvesoftware.Steam.CompatibilityTool.Boxtron`, and change other things like the project name in appdata when it's actually renamed.

**Comment by dreamer** [comment]
Yes, the name Boxtron is final - I am holding off with the rename because I need to test beforehand if GitHub will correctly perform redirects to the new name in all contexts (wiki links, forks, Travis, etc). I think this extension ID is good :)

**Comment by gasinvein** [comment]
@nanonyme Are there any plans to merge the extension support into stable branch?

**Comment by gasinvein** [comment]
Manifest with the new ID is now [here](https://github.com/gasinvein/flatpak-apps-staging/commits/com.valvesoftware.Steam.CompatibilityTool.Boxtron/master). Can confirm that it keeps working.

**Comment by nanonyme** [comment]
I'm intending to just rebase master against beta once 18.08 is released. It's date-based versioning so should be out around August to September. Beta is going to start picking up 32bit versions of bundled libs once I manage to fix build.

**Comment by nanonyme** [comment]
Luxtorpeda actually sounds harder if it depends on Rust. Or is onlly 64bit okay for that?

**Comment by dreamer** [comment]
Why Rust would be a problem? It's only a build-time dependency :) I presume Flatpak runtime already provides all runtime dependencies. Release binary built on Fedora 30:

    $ ldd target/release/luxtorpeda
    	linux-vdso.so.1 (0x00007ffc96bf3000)
    	libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007f65a03a2000)
    	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007f65a00c1000)
    	libdl.so.2 => /lib64/libdl.so.2 (0x00007f65a00bb000)
    	librt.so.1 => /lib64/librt.so.1 (0x00007f65a00b1000)
    	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f65a0090000)
    	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f65a0076000)
    	libc.so.6 => /lib64/libc.so.6 (0x00007f659feae000)
    	/lib64/ld-linux-x86-64.so.2 (0x00007f65a0924000)
    	libm.so.6 => /lib64/libm.so.6 (0x00007f659fd68000)
    	libz.so.1 => /lib64/libz.so.1 (0x00007f659fd4e000)

I do not intend to support 32-bit (according to several different surveys, number of Linux gamers using 32-bit distributions is ~0% :) ). All future builds of Luxtorpeda will be exclusively `x86_64`. I don't know if this will conflict somehow with existing Steam flatpak package (I presume not).

As for packages distributed through Luxtorpeda - they depend on Steam Runtime, I also made specific choice to support 64-bit only.

**Comment by gasinvein** [comment]
@dreamer Rust build-time is exactly the potential source of problems, since Flathub's infra prohibits network access during build. However, there are [tools](https://github.com/flatpak/flatpak-builder-tools) to overcome this.

**Comment by dreamer** [comment]
Hmm, interesting… I googled and read through various issues that were opened/closed regarding this topic and it seems like there's no clear, generic solution yet…

What if I distributed a special release tarball, that would include source code for all the cargo package dependencies exactly in versions required to build the project? Flatpak infrastructure could use that tarball to populate cargo cache and perform the build completely offline. If this approach would work, we could turn it into a new cargo command or option (`cargo package --include-dependencies`?).

**Comment by gasinvein** [comment]
@dreamer Yes, vendored sources (archive with dependencies) should always work, but it's not necessarily necessary. The `flatpak-cargo-generator` generates a drop-in that makes `flatpak-builder` download cargo crates during prepare stage, so build stage can successfully run offline.

**Comment by nanonyme** [comment]
@dreamer sounds like you are confused by the survey results. That around 0% is users dealing with 32bit kernel and 32bit-only userland. 100% of Steam users have 32bit userland though multilib support through compat libs. Is this thing relevant to any 32bit only games?

**Comment by dreamer** [comment]
@nanonyme you are right, but I am not confused. There is simply no need for my project to be on 32-bit only system. I rebuild open source game engines to work on 64-bit steam runtime. There are no users who would benefit from me compiling Luxtorpeda (or Luxtorpeda packages) for 32-bit. And it makes everyone's life easier :)

**Comment by nanonyme** [comment]
That's different. Then there's only the Rust part remaining. It's not part of runtime so you need to do your stuff with the Sdk extension. It's only available during build-time by default, not runtime.

**Comment by gasinvein** [comment]
Luxtorpeda flatpak builds just fine, have a look at the manifest [here](https://github.com/gasinvein/flatpak-apps-staging/tree/com.valvesoftware.Steam.CompatibilityTool.Luxtorpeda/master). Didn't actually try running it since I don't have any of it's supported games.

**Comment by dreamer** [comment]
@gasinvein do you own any of the games mentioned at [wiki](https://github.com/dreamer/luxtorpeda/wiki/Game-engines)? Maybe I can prioritize package release for a specific title. If not, there's also an option of testing [DDLC](https://store.steampowered.com/app/698780/Doki_Doki_Literature_Club/) - it's a free game with Linux version bundled inside Windows depot, so Luxtorpeda simply rewrites the command used to start the game. If it starts natively (not through Proton), it means Luxtorpeda is working correctly.

**Comment by gasinvein** [comment]
@dreamer
>do you own any of the games mentioned at wiki?

 Only AoE II, but openage [doesn't work](https://github.com/SFTtech/openage/issues/788) with the latest AoE II addon.

>is it possible with Flatpak 1.4.2?

Sure, why wouldn't it?

**Comment by dreamer** [comment]
Ok, so how can I install packages, that are in staging? Or should I just wait some time for package to appear in `flatpak search` results?

**Comment by gasinvein** [comment]
@dreamer flatpak-apps-staging is just the name of my git repo, it's not related to Flathub. You could build apps from there yourself, locally. You would only need the "flathub-beta" [repo](https://flathub.org/beta-repo/flathub-beta.flatpakrepo) added, `flatpak-builder` should do the rest.

**Comment by AsciiWolf** [comment]
Any news regarding the  [Luxtorpeda](https://github.com/luxtorpeda-dev/luxtorpeda) CompatibilityTool Steam Flatpak extension? :-)

**Comment by gasinvein** [comment]
@AsciiWolf Hm, I thought its development stopped 'cause there is no activity in the original repo. Looks like it has a new maintainer now?

**Comment by AsciiWolf** [comment]
Yeah, seems so.

**Comment by gasinvein** [comment]
@AsciiWolf Probably you should ask their opinion on submitting it to Flathub, then.

**Comment by AsciiWolf** [comment]
@gasinvein I have made a RFE ticket in the luxtorpeda repo.

**Comment by d10sfan** [comment]
Hello, 

I'm the maintainer of luxtorpeda-dev and saw the ticket about adding it as a compatibility tool to flathub. I'm not very familiar with submitting packages or at least tools for flatpack steam, so is there some documentation or pointers I could look at?

I saw this: https://github.com/gasinvein/flatpak-apps-staging/blob/com.valvesoftware.Steam.CompatibilityTool.Luxtorpeda/beta/com.valvesoftware.Steam.CompatibilityTool.Luxtorpeda.yml

Which looks like it'd be a good start, as the luxtorpeda client does still use rust, although some of the packages have changed.

**Comment by gasinvein** [comment]
@d10sfan Common complication with building Rust apps on Flathub is that they must be built offline, i.e. cargo won't be able to pull crates from network. There are two ways around this: use tarball with vendored sources (if upstream provides one) or use the [flatpak-cargo-generator.py](https://github.com/flatpak/flatpak-builder-tools/tree/master/cargo) helper script that reads `Cargo.lock` and produces a drop-in manifest with crates (the `generated-sources.json` thing).
Other than this, building single Rust binary should be pretty straightforward.
I've updated the manifest switching to your repo, Luxtorpeda still builds fine and Steam is able to find it, but I have no idea how to use/test it.

**Comment by dreamer** [comment]
As for packaging Luxtorpeda (either official or dev fork, which might soon merge back into a single repo):

1. Luxtorpeda (my original repo) used only Steam Runtime, therefore it theoretically could be packaged via Flatpak, however I *never released this software as a stable version*. I very much intend to do backwards-incompatible, radical changes to the tool.
2. Luxtorpeda-dev fixed the issues we had with packaging certain games (some games do really terrible job when it comes to managing their dependencies) - but in order to do so, it started using libraries from host OS (so games depending on Ubuntu 18.04 libraries likely won't work via Flatpak).
3. Since Luxtorpeda is basically a **package manager** (I don't think Flatpak is designed to be used in this case), I was considering turning the tool into a Flatpak runtime. I did no research towards this goal yet, it entirely depends on what options are provided to us via new Steam Runtime.

**Comment by gasinvein** [comment]
@dreamer I believe the new Steam Runtime has basically the same limitations as flatpak in that regard - nothing from host OS (except for GL drivers), everything must be either present in the runtime or bundled.

**Comment by d10sfan** [comment]
One thing to note for the ubuntu 18.04 packages games is that it still provides libraries and does not rely on the host OS libraries, so those should still work ok, but haven't tested those with flatpak before. Ubuntu 18.04 is mostly used as a base for the build process, but it isn't expecting those libraries to be supported by the host. Also there are still many engines that work with the normal steam runtime.

I saw this PR come in: https://github.com/luxtorpeda-dev/luxtorpeda/pull/55 Is this related to this thread at all?

**Comment by gasinvein** [comment]
@d10sfan Just to clarify, do you build you game engines against the steam runtime (the old one) or against Ubuntu 18.04?
>I saw this PR come in: luxtorpeda-dev/luxtorpeda#55 Is this related to this thread at all?

I guess so. [AppStream metadata](https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html) in general is not flatpak-specific and is useful for rpm/deb repositories as well, but the file in that PR probably is flatpak-specific since it uses IDs specific to this (Steam) flatpak.

**Comment by d10sfan** [comment]
luxtorpeda-dev uses a combination of both the steam runtime and ubuntu 18.04

For any engine that can support steam runtime, it'll use that for building, and for any extra libraries needed on top of what steam runtime provides it'll also provide those in the engine build artifact.

For an engine that needs Ubuntu 18.04, usually as a result of a glibc upgrade need or something similar, the build process takes a bare-bones ubuntu:18.04 docker image and then will build off of that. 

The ubuntu 18.04 image only gets these following packages installed directly from apt-get. The rest are provided by building the libraries and providing them in the build artifact. Those libraries that are included in the build artifact could include SDL2, openal, etc.

apt-get -y install build-essential cmake git g++-8 gcc-8 sudo wget unzip libx11-dev libgl1-mesa-dev automake libtool ncurses-dev pkg-config libpulse-dev freeglut3-dev libxrandr-dev libxinerama-dev libasound2-dev libxcursor-dev libxi-dev

wget, unzip, and the build dependencies are only needed by the build itself and not needed by the built engine. So as long as the included graphics libraries include these x11 dependencies, then it should be ok. If not, then let me know and I can look into providing those as well.

**Comment by gasinvein** [comment]
>build-essential cmake git g++-8 gcc-8 sudo wget unzip libx11-dev libgl1-mesa-dev automake libtool ncurses-dev pkg-config libpulse-dev freeglut3-dev libxrandr-dev libxinerama-dev libasound2-dev libxcursor-dev libxi-dev

I guess freedesktop-sdk (the flatpak runtime) has everything from the list exept for freeglut, but @nanonyme should know better.
Yet we'll have to test ABI compatibility, I guess. E.g. will something built against libpulse from Ubuntu work with libpulse from the flatpak runtime?

**Comment by nanonyme** [comment]
I'm doubtful about libxinerama, libxcursor and libxi.

**Comment by dreamer** [comment]
> Yet we'll have to test ABI compatibility, I guess.

I don't think you should - Luxtorpeda-provided packages need to have a stable base and maintainable, future-proof build environment first.

> I'm doubtful about libxinerama, libxcursor and libxi.

Exactly. And I am worried about using Ubuntu 18.04 as a base in general, because of ABI-incompatibilities resulting from usage of C++ programs and libraries compiled with different compilers/compiler versions. Valve developers allow, but explicitly do not support running games via native runtimes (which is a sane and proper decision on their part).

To put it more bluntly: in Luxtorpeda I don't want to have a situation when C++ game is compiled with compiler A and depend at the same time on C++ libraries provided by Steam (compiled with compiler B) and C++ libraries provided by another runtime (compiled with compiler C). Right now it "usually works", but it's already a wonky solution. I predict the situation will turn to worse once game developers will start jumping onto C++20.

My opinion here is: *We haven't explored the problem space to know how to design the solution that's both maintainable and future-proof. Luxtorpeda is proof-of-concept, not a finished solution.*

Therefore I advise to not package Luxtorpeda nor Luxtorpeda-dev in Flatpak (yet)…

**Comment by nanonyme** [comment]
Sorry but where does thus Ubuntu 18.04 thing come from? Freedesktop-sdk is a standalone thing and Steam Runtime is based on Ubuntu 12.04

**Comment by nanonyme** [comment]
I would suggest versioning extension by freedesktop-sdk (we should do that everywhere really but haven't) and not build anything outside flatpak (eg in Docker). It is unlikely to end well.

**Comment by gasinvein** [comment]
>Luxtorpeda-provided packages need to have a stable base and maintainable, future-proof build environment first.

Hm, did you consider using freedesktop-sdk? It has stable ABI and known set of libraries during one version lifecycle. It's available as docker images, so you don't need flatpak to build against it.

**Comment by d10sfan** [comment]
Does the freedesktop sdk help with non flatpak environments, such as running steam directly on the host?

I believe one research item would be to look at steam's new runtime although I believe it currently has compatibility issues with running inside flatpak

**Comment by gasinvein** [comment]
@d10sfan It depends on if it would be more compatible with non-ubuntu distros than Ubuntu. Personally I think it could be.

**Comment by dreamer** [comment]
> Sorry but where does thus Ubuntu 18.04 thing come from? Freedesktop-sdk is a standalone thing and Steam Runtime is based on Ubuntu 12.04

@nanonyme Luxtorpeda is package manager installing 3rd-party game engines to work with Steam-supplied game data. The client program is written in Rust and can be relatively easily packaged on FlatHub (that's not an issue). However the games themselves are not built inside Flatpak - they depend on old Steam Runtime (Luxtorpeda), or Steam Runtime+Ubuntu 18.04 runtime (Luxtorpeda-dev) and are built inside Docker containers provided by us.

> Hm, did you consider using freedesktop-sdk? It has stable ABI and known set of libraries during one branch lifecycle. It's available as docker images, so you don't need flatpak to build against it.

Yes. It would work for some games (mostly those that we can easily run in pure Steam Runtime environment now anyway), but it would **not** work from maintainer's perspective (rebuilding all dependencies required for *some* games can easily turn into multi-hour affairs, that can easily timeout on various free-tier CI solutions).

I am considering building Luxtorpeda-specific flatpak runtime(s) and distributing games via flatpak (instead of our own format), which would turn Luxtorpeda client into a thin layer of glue connecting Steam with Flatpak-packaged games. This would provide several upsides, but I haven't started work on this yet (nor I haven't decided this is definitely the solution I will choose).

**Comment by d10sfan** [comment]
Ill also be doing some research to see if it's possible to use steam's new container based runtime

**Comment by nanonyme** [comment]
It will not work with flatpak.

**Comment by d10sfan** [comment]
Out of curiosity, does Proton 5.13 and/or proton-ge (5.13 and above) also not work with flatpak?

**Comment by gasinvein** [comment]
@d10sfan Yes, anything using the new steam runtime won't work in flatpak, since it uses its own app container, pressure-vessel. See #642 for more details.

**Comment by AsciiWolf** [comment]
It will hopefully work in the future, as far as I know, but we are not there yet.

**Comment by AsciiWolf** [comment]
I wonder if it could be possible to add Luxtorpeda at least to the Flathub beta branch. :-)

**Comment by AsciiWolf** [comment]
Sorry for pinging, but I wonder if there is any progress in getting Luxtorpeda working with Flatpak Steam.

**Comment by gasinvein** [comment]
@AsciiWolf Did you try installing it manually, i.e. to `~/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d`?

**Comment by AsciiWolf** [comment]
Nope, not yet. I will try it.

**Comment by jkonecny12** [comment]
If I understand this comment correctly https://github.com/flathub/com.valvesoftware.Steam/issues/642#issuecomment-830182456 than the issue with app-container in flatpak should be solved by Flatpak 1.11.2 version?

**Comment by gasinvein** [comment]
@jkonecny12 Yes, Valve's container runner, pressure-vessel, now works in flatpak 1.11+ sandbox, thanks to @smcv.

**Comment by nanonyme** [comment]
I'm keeping this open and pinned at least until 1.12.0 is out.

**Comment by nanonyme** [comment]
That's not to say shipping unofficial Proton forks etc might still not be justified

**Comment by gasinvein** [comment]
I guess the main question in this thread is if Luxtorpeda devs should build it and its runners against Steam runtime container.

**Comment by smcv** [comment]
If Luxtorpeda is currently using its own ad-hoc environment involving Ubuntu 18.04 + addons, then Steam Runtime 2 'soldier' might be a good replacement for that. With my Steam Runtime hat on, I need to maintain Steam Runtime 2 'soldier' for Proton *anyway*...

I'm hoping that in future, Steam can get an official way to run native Linux games under Steam Runtime 2 'soldier' (based on Debian 10) or the as-yet-unreleased Steam Runtime 3 'sniper' (based on Debian 11), but the necessary infrastructure inside the proprietary Steam client isn't there yet. At the moment, Steam effectively always assumes that native Linux games are intended to run on Steam Runtime 1 'scout', based on Ubuntu 12.04, and there's currently no mechanism to flag individual games with "actually, this one would prefer Steam Runtime 2".

I can't promise anything about long-term support for soldier, but a lot of the reason we're using a Debian derivative instead of the fd.o SDK is to avoid having to migrate to a new release every year - Debian stable releases are approximately 1 per 2 years, and have overlapping support lifetimes. soldier is based on Debian 10, so it's about a year newer than Ubuntu 18.04, and I expect that it'll pick up periodic security updates from Debian until mid 2022, then from Debian LTS (probably less frequently) until mid 2024. After that, I don't know.

We actually did try doing an Ubuntu-18.04-based runtime, 'spy', before soldier - but being a derivative-of-a-derivative makes it harder to rebase changes, so we moved to the Debian-based soldier instead of ever releasing spy.

**Comment by d10sfan** [comment]
@smcv I'm the maintainer of luxtorpeda-dev, having a way to use the solider or sniper runtime would be great, both for the building of the engines and the execution of them on the steam client side.

Is there a way with the right commands to execute the solider runtime using the command line (assuming it's already been installed). If so, I could have a way of doing the same using luxtorpeda by having it run that command to launch the game, assuming it's not too late.

**Comment by gasinvein** [comment]
> If so, I could have a way of doing the same using luxtorpeda by having it run that command to launch the game, assuming it's not too late.

I _think_ you can declare the steam runtime you want to use in your `toolmanifest.vdf` file, just like Proton does; then I _think_ Steam will handle launching luxtorpeda with proper container runtime for you.

**Comment by smcv** [comment]
> I _think_ you can declare the steam runtime you want to use in your `toolmanifest.vdf` file, just like Proton does

Yes, you can "stack" compatibility tools, and that's how Proton 5.13+ is set up to run on soldier - there's no other magic involved. [Proton-GE also uses this technique](https://github.com/GloriousEggroll/proton-ge-custom/blob/master/toolmanifest_runtime.vdf).

I'm trying to keep https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/blob/master/docs/steam-compat-tool-interface.md up to date with the way compatibility tools work inside Steam; but I don't have access to Steam source code, so that document cannot be completely authoritative.

> for the building of the engines

We have an official Docker image, documented at https://gitlab.steamos.cloud/steamrt/soldier/sdk (the short version is `sudo docker pull registry.gitlab.steamos.cloud/steamrt/soldier/sdk`).

Or, you can grab `com.valvesoftware.SteamRuntime.Sdk-amd64,i386-soldier-sysroot.Dockerfile` and `com.valvesoftware.SteamRuntime.Sdk-amd64,i386-soldier-sysroot.tar.gz` from https://repo.steampowered.com/steamrt-images-soldier/snapshots/latest-container-runtime-depot/ (or any older version) and `docker build` it yourself, if you prefer to do it that way.

sniper is not available to the public yet, but when it becomes available it's very likely to be the same "shape" as soldier, just with `soldier` replaced by `sniper` everywhere.

**Comment by d10sfan** [comment]
Thanks I'll take a look at that

**Comment by gasinvein** [comment]
Note that if you build Luxtorpeda against Steam runtime, you won't be able to build Luxtorpeda on Flathub.
Unless... Can we have `com.valvesoftware.SteamRuntime.Sdk` on Flathub?

**Comment by d10sfan** [comment]
I thought that the issues with flatpak steam and the new runtime were fixed in the development versions?

**Comment by gasinvein** [comment]
> I thought that the issues with flatpak steam and the new runtime were fixed in the development versions?

They are. My post was about how you build and publish Luxtorpeda, not about how users run it. With flatpak 1.11+, users will be able to run Luxtorpeda in steam runtime container, but you won't be able to build it from source on Flathub infra.

**Comment by d10sfan** [comment]
My main plan was about changing the way the engines are built to use the new runtime, not luxtorpeda client itself. The luxtorpeda client should all have deps built-in, so shouldn't need any changes on how it is built.

**Comment by AsciiWolf** [comment]
It [looks like](https://www.gamingonlinux.com/2021/10/steam-play-tool-luxtorpeda-for-running-games-in-native-linux-engines-sees-a-major-upgrade) that Luxtorpeda has switched to the new Steam Runtime (Soldier). :-)
